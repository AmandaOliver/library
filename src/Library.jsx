import React, { PureComponent } from 'react'
import Header from './Header'
import Catalog from './Catalog'
import { Provider } from './context'
import {sortArrayBooks, filterArrayBooks, getBookById} from './libraryLoader'
import config from './configuration'

class Library extends PureComponent {
  constructor() {
    super()

    this.actions = {
      getLibrarySize: this._getLibrarySize,
      getBookById: this._getBookById,
      sortByProperty: this._sortByProperty,
      filterBooks: this._filterBooks,
      resetFilter: this._resetFilter,
    }

    this.state = {
      filteredBooksArray: [],
      isFiltered: false,
      sortedBy: 'id',
      ...this.actions,
    }

  }
  _resetFilter = () => {
    this.setState({
      filteredBooksArray: [],
      isFiltered: false
    })
  }
  _getLibrarySize = () => this.state.isFiltered ? this.state.filteredBooksArray.length : config.librarySize
  _getBookById = id => this.state.isFiltered ? this.state.filteredBooksArray[id] : getBookById(id)
  _sortByProperty = property => {
    const sortBooks = (property, books) => {
      if (property === 'id') {
        books.sort((bookA, bookB) => bookA.id - bookB.id)
      }
      books.sort((bookA, bookB) => bookA.bookData[property] < bookB.bookData[property] ? -1 : 1)
    }
    if (this.state.isFiltered) {
      sortBooks(property, this.state.filteredBooksArray)
    }
    sortArrayBooks(property)
    this.setState({sortedBy: property})
  }

  _filterBooks = (bookProperties) => {
    this.setState({
      filteredBooksArray: filterArrayBooks(bookProperties),
      isFiltered: true
    })
  }

  render() {
    return (
      <Provider value={this.state}>
        <Header />
        <Catalog />
      </Provider>
    )
  }
}

export default Library
