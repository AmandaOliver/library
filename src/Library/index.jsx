import React, { PureComponent } from 'react'
import Header from '../Header'
import Catalog from '../Catalog'
import { Provider } from '../context'
import lib from '../libraryLoader'
import config from '../configuration'
import { sortBooks } from '../utils'

class Library extends PureComponent {
  constructor() {
    super()

    this.actions = {
      getLibrarySize: this._getLibrarySize,
      getBookById: this._getBookById,
      sortByProperty: this._sortByProperty,
      filterByProperties: this._filterByProperties,
      resetFilter: this._resetFilter,
    }

    this.state = {
      filteredBooksArray: [],
      isFiltered: false,
      sortedBy: 'id',
      ...this.actions,
    }

  }
  _resetFilter = () => {
    this.setState({
      filteredBooksArray: [],
      isFiltered: false
    })
  }
  _getLibrarySize = () => this.state.isFiltered ? this.state.filteredBooksArray.length : config.librarySize
  _getBookById = id => this.state.isFiltered ? this.state.filteredBooksArray[id] : lib.getBookById(id)
  _sortByProperty = property => {
    if (this.state.isFiltered) {
      sortBooks(property, this.state.filteredBooksArray)
    }
    lib.sortByProperty(property)
    this.setState({sortedBy: property})
  }

  _filterByProperties = (paramsArray) => {
    this.setState({
      filteredBooksArray: lib.filterByProperties(paramsArray),
      isFiltered: true
    })
  }

  render() {
    return (
      <Provider value={this.state}>
        <Header />
        <Catalog />
      </Provider>
    )
  }
}

export default Library
