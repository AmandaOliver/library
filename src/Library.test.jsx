import React from 'react'
import renderer from 'react-test-renderer'
import Library from './Library'

jest.mock('./libraryGenerator.worker', () =>
  class Worker {
    constructor() {this.onmessage = () => {}}
    addEventListener() { return }
    postMessage(msg) {
      this.onmessage(msg)
    }
  }
)
describe('Library', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <Library />,
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
