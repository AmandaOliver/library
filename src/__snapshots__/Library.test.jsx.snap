// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library matches snapshot 1`] = `
Array [
  <header
    className="header"
  >
    <span
      className="header__title"
    >
      THE LIBRARY
    </span>
    <div
      className="header__loading"
    >
      Loading
    </div>
  </header>,
  <div
    className="catalog"
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    >
      <div
        aria-label="grid"
        aria-readonly={true}
        className="ReactVirtualized__Grid ReactVirtualized__List"
        onScroll={[Function]}
        role="grid"
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "boxSizing": "border-box",
            "direction": "ltr",
            "height": 0,
            "overflowX": "hidden",
            "overflowY": "auto",
            "position": "relative",
            "width": 0,
            "willChange": "transform",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>,
]
`;
